common/utils/include/claragenomics/utils/cudautils.hpp

```diff
@@ -43,24 +43,28 @@ namespace cudautils
 /// \param code The CUDA status code of the function being asserted
 /// \param file Filename of the calling function
 /// \param line File line number of the calling function
-/// \param abort If true, hard-exit on CUDA error
-inline void gpu_assert(cudaError_t code, const char* file, int line, bool abort = false)
+/// \param cuda_sync Synchronize CUDA device before checking error
+inline void gpu_assert(cudaError_t code, const char* file, int line, bool cuda_sync = false)
 {
+    // This device synchronize forces the most recent CUDA call to fully
+    // complete, increasing the chance of catching the CUDA error near the
+    // offending function.
+    if (cuda_sync)
```

Comment:
`cudaDeviceSynchronize()` could potentially overwrite a previous `code` value. I think changing this if to `cuda_sync && code == cudaSuccess)` should fix the problem.

Reply:
Wouldn't the last bad error code stick for `cudaDeviceSynchronize()`? I always assumed that was the behavior of that call, and that error states in the driver couldn't be overwritten. But for safety, let me change it.

Reply:
The more serious errors will, but if you get an out-of-memory from a cudaMalloc for example, it will not reappear when you call `cudaDeviceSynchronize()`.

Reply:
Hmm good point, that would indeed be nasty if an our of memory error got undetected. Pushed the changed